File: frontend/src/utils/validators.ts
/**
 * Utilitários de validação para formulários
 */

/**
 * Remove caracteres não numéricos de uma string
 */
export const removeNonNumeric = (value: string): string => {
  return value.replace(/\D/g, "");
};

/**
 * Valida e formata CPF
 */
export const validateCPF = (
  cpf: string,
): { isValid: boolean; formatted: string; error?: string } => {
  const cleanCPF = removeNonNumeric(cpf);

  if (cleanCPF.length !== 11) {
    return { isValid: false, formatted: cpf, error: "CPF deve ter 11 dígitos" };
  }

  // Verifica se todos os dígitos são iguais
  if (/^(\d)\1{10}$/.test(cleanCPF)) {
    return { isValid: false, formatted: cpf, error: "CPF inválido" };
  }

  // Validação dos dígitos verificadores
  let sum = 0;
  for (let i = 0; i < 9; i++) {
    sum += parseInt(cleanCPF.charAt(i)) * (10 - i);
  }
  let remainder = (sum * 10) % 11;
  if (remainder === 10 || remainder === 11) remainder = 0;
  if (remainder !== parseInt(cleanCPF.charAt(9))) {
    return { isValid: false, formatted: cpf, error: "CPF inválido" };
  }

  sum = 0;
  for (let i = 0; i < 10; i++) {
    sum += parseInt(cleanCPF.charAt(i)) * (11 - i);
  }
  remainder = (sum * 10) % 11;
  if (remainder === 10 || remainder === 11) remainder = 0;
  if (remainder !== parseInt(cleanCPF.charAt(10))) {
    return { isValid: false, formatted: cpf, error: "CPF inválido" };
  }

  const formatted = `${cleanCPF.slice(0, 3)}.${cleanCPF.slice(3, 6)}.${cleanCPF.slice(6, 9)}-${cleanCPF.slice(9)}`;
  return { isValid: true, formatted };
};

/**
 * Formata CPF enquanto o usuário digita
 */
export const formatCPF = (value: string): string => {
  const cleanValue = removeNonNumeric(value);
  if (cleanValue.length <= 3) return cleanValue;
  if (cleanValue.length <= 6)
    return `${cleanValue.slice(0, 3)}.${cleanValue.slice(3)}`;
  if (cleanValue.length <= 9)
    return `${cleanValue.slice(0, 3)}.${cleanValue.slice(3, 6)}.${cleanValue.slice(6)}`;
  return `${cleanValue.slice(0, 3)}.${cleanValue.slice(3, 6)}.${cleanValue.slice(6, 9)}-${cleanValue.slice(9, 11)}`;
};

/**
 * Valida e formata telefone brasileiro
 */
export const validatePhone = (
  phone: string,
): { isValid: boolean; formatted: string; error?: string } => {
  const cleanPhone = removeNonNumeric(phone);

  if (cleanPhone.length < 10 || cleanPhone.length > 11) {
    return {
      isValid: false,
      formatted: phone,
      error: "Telefone deve ter 10 ou 11 dígitos",
    };
  }

  // Verifica se é celular (11 dígitos) ou fixo (10 dígitos)
  let formatted: string;
  if (cleanPhone.length === 11) {
    // Celular: (XX) 9XXXX-XXXX
    formatted = `(${cleanPhone.slice(0, 2)}) ${cleanPhone.slice(2, 7)}-${cleanPhone.slice(7)}`;
  } else {
    // Fixo: (XX) XXXX-XXXX
    formatted = `(${cleanPhone.slice(0, 2)}) ${cleanPhone.slice(2, 6)}-${cleanPhone.slice(6)}`;
  }

  return { isValid: true, formatted };
};

/**
 * Formata telefone enquanto o usuário digita
 */
export const formatPhone = (value: string): string => {
  const cleanValue = removeNonNumeric(value);
  if (cleanValue.length <= 2) return cleanValue;
  if (cleanValue.length <= 6)
    return `(${cleanValue.slice(0, 2)}) ${cleanValue.slice(2)}`;
  if (cleanValue.length <= 10)
    return `(${cleanValue.slice(0, 2)}) ${cleanValue.slice(2, 6)}-${cleanValue.slice(6)}`;
  return `(${cleanValue.slice(0, 2)}) ${cleanValue.slice(2, 7)}-${cleanValue.slice(7, 11)}`;
};

/**
 * Valida e formata CEP
 */
export const validateCEP = (
  cep: string,
): { isValid: boolean; formatted: string; error?: string } => {
  const cleanCEP = removeNonNumeric(cep);

  if (cleanCEP.length !== 8) {
    return { isValid: false, formatted: cep, error: "CEP deve ter 8 dígitos" };
  }

  const formatted = `${cleanCEP.slice(0, 5)}-${cleanCEP.slice(5)}`;
  return { isValid: true, formatted };
};

/**
 * Formata CEP enquanto o usuário digita
 */
export const formatCEP = (value: string): string => {
  const cleanValue = removeNonNumeric(value);
  if (cleanValue.length <= 5) return cleanValue;
  return `${cleanValue.slice(0, 5)}-${cleanValue.slice(5, 8)}`;
};

/**
 * Valida data de nascimento
 */
export const validateBirthDate = (
  date: string,
): { isValid: boolean; error?: string } => {
  if (!date) {
    return { isValid: false, error: "Data de nascimento é obrigatória" };
  }

  const birthDate = new Date(date);
  const today = new Date();

  if (isNaN(birthDate.getTime())) {
    return { isValid: false, error: "Data inválida" };
  }

  if (birthDate >= today) {
    return {
      isValid: false,
      error: "Data de nascimento deve ser anterior à data atual",
    };
  }

  // Calcular idade
  const age = today.getFullYear() - birthDate.getFullYear();
  const monthDiff = today.getMonth() - birthDate.getMonth();
  const dayDiff = today.getDate() - birthDate.getDate();

  const actualAge =
    monthDiff < 0 || (monthDiff === 0 && dayDiff < 0) ? age - 1 : age;

  if (actualAge < 16) {
    return { isValid: false, error: "Idade mínima é 16 anos" };
  }

  return { isValid: true };
};

/**
 * Valida estado brasileiro
 */
export const validateState = (
  state: string,
): { isValid: boolean; error?: string } => {
  const validStates = [
    "AC",
    "AL",
    "AP",
    "AM",
    "BA",
    "CE",
    "DF",
    "ES",
    "GO",
    "MA",
    "MT",
    "MS",
    "MG",
    "PA",
    "PB",
    "PR",
    "PE",
    "PI",
    "RJ",
    "RN",
    "RS",
    "RO",
    "RR",
    "SC",
    "SP",
    "SE",
    "TO",
  ];

  if (!validStates.includes(state.toUpperCase())) {
    return { isValid: false, error: "Estado inválido" };
  }

  return { isValid: true };
};

/**
 * Lista de estados brasileiros para select
 */
export const BRAZILIAN_STATES = [
  { value: "AC", label: "Acre" },
  { value: "AL", label: "Alagoas" },
  { value: "AP", label: "Amapá" },
  { value: "AM", label: "Amazonas" },
  { value: "BA", label: "Bahia" },
  { value: "CE", label: "Ceará" },
  { value: "DF", label: "Distrito Federal" },
  { value: "ES", label: "Espírito Santo" },
  { value: "GO", label: "Goiás" },
  { value: "MA", label: "Maranhão" },
  { value: "MT", label: "Mato Grosso" },
  { value: "MS", label: "Mato Grosso do Sul" },
  { value: "MG", label: "Minas Gerais" },
  { value: "PA", label: "Pará" },
  { value: "PB", label: "Paraíba" },
  { value: "PR", label: "Paraná" },
  { value: "PE", label: "Pernambuco" },
  { value: "PI", label: "Piauí" },
  { value: "RJ", label: "Rio de Janeiro" },
  { value: "RN", label: "Rio Grande do Norte" },
  { value: "RS", label: "Rio Grande do Sul" },
  { value: "RO", label: "Rondônia" },
  { value: "RR", label: "Roraima" },
  { value: "SC", label: "Santa Catarina" },
  { value: "SP", label: "São Paulo" },
  { value: "SE", label: "Sergipe" },
  { value: "TO", label: "Tocantins" },
];

File: backend/app/utils/validators.py
"""Utilitários de validação para campos do usuário."""

import re
from datetime import date, datetime
from typing import Union


def validate_cpf(cpf: str) -> bool:
    """Valida CPF brasileiro.

    Args:
        cpf: CPF a ser validado (com ou sem formatação)

    Returns:
        bool: True se CPF é válido, False caso contrário
    """
    if not cpf:
        return False

    # Remove formatação
    cpf = re.sub(r"[^0-9]", "", cpf)

    # Verifica se tem 11 dígitos
    if len(cpf) != 11:
        return False

    # Verifica se todos os dígitos são iguais
    if cpf == cpf[0] * 11:
        return False

    # Calcula primeiro dígito verificador
    soma = sum(int(cpf[i]) * (10 - i) for i in range(9))
    resto = soma % 11
    digito1 = 0 if resto < 2 else 11 - resto

    # Verifica primeiro dígito
    if int(cpf[9]) != digito1:
        return False

    # Calcula segundo dígito verificador
    soma = sum(int(cpf[i]) * (11 - i) for i in range(10))
    resto = soma % 11
    digito2 = 0 if resto < 2 else 11 - resto

    # Verifica segundo dígito
    return int(cpf[10]) == digito2


def validate_phone(phone: str) -> bool:
    """Valida telefone brasileiro.

    Args:
        phone: Telefone a ser validado

    Returns:
        bool: True se telefone é válido, False caso contrário
    """
    if not phone:
        return False

    # Remove formatação
    phone = re.sub(r"[^0-9]", "", phone)

    # Verifica se tem 10 ou 11 dígitos (com DDD)
    if len(phone) not in [10, 11]:
        return False

    # Verifica se DDD é válido (11 a 99)
    ddd = int(phone[:2])
    if ddd < 11 or ddd > 99:
        return False

    # Para celular (11 dígitos), o terceiro dígito deve ser 9
    if len(phone) == 11 and phone[2] != "9":
        return False

    return True


def validate_cep(cep: str) -> bool:
    """Valida CEP brasileiro.

    Args:
        cep: CEP a ser validado

    Returns:
        bool: True se CEP é válido, False caso contrário
    """
    if not cep:
        return False

    # Remove formatação
    cep = re.sub(r"[^0-9]", "", cep)

    # Verifica se tem 8 dígitos
    return len(cep) == 8 and cep.isdigit()


def validate_birth_date(birth_date: Union[date, str]) -> bool:
    """Valida data de nascimento (idade mínima de 16 anos).

    Args:
        birth_date: Data de nascimento

    Returns:
        bool: True se data é válida, False caso contrário
    """
    if not birth_date:
        return False

    try:
        if isinstance(birth_date, str):
            # Tenta converter string para date
            birth_date = datetime.strptime(birth_date, "%Y-%m-%d").date()

        # Verifica se é uma data válida
        if not isinstance(birth_date, date):
            return False

        # Verifica se não é uma data futura
        today = date.today()
        if birth_date > today:
            return False

        # Calcula idade
        age = today.year - birth_date.year
        if today.month < birth_date.month or (
            today.month == birth_date.month and today.day < birth_date.day
        ):
            age -= 1

        # Verifica idade mínima de 16 anos
        return age >= 16

    except (ValueError, TypeError):
        return False


def validate_state(state: str) -> bool:
    """Valida sigla do estado brasileiro.

    Args:
        state: Sigla do estado (2 caracteres)

    Returns:
        bool: True se estado é válido, False caso contrário
    """
    if not state:
        return False

    # Lista de estados brasileiros válidos
    valid_states = {
        "AC",
        "AL",
        "AP",
        "AM",
        "BA",
        "CE",
        "DF",
        "ES",
        "GO",
        "MA",
        "MT",
        "MS",
        "MG",
        "PA",
        "PB",
        "PR",
        "PE",
        "PI",
        "RJ",
        "RN",
        "RS",
        "RO",
        "RR",
        "SC",
        "SP",
        "SE",
        "TO",
    }

    return state.upper() in valid_states


def format_cpf(cpf: str) -> str:
    """Formata CPF para exibição.

    Args:
        cpf: CPF sem formatação

    Returns:
        str: CPF formatado (XXX.XXX.XXX-XX)
    """
    if not cpf:
        return ""

    # Remove formatação existente
    cpf = re.sub(r"[^0-9]", "", cpf)

    if len(cpf) == 11:
        return f"{cpf[:3]}.{cpf[3:6]}.{cpf[6:9]}-{cpf[9:]}"

    return cpf


def format_phone(phone: str) -> str:
    """Formata telefone para exibição.

    Args:
        phone: Telefone sem formatação

    Returns:
        str: Telefone formatado
    """
    if not phone:
        return ""

    # Remove formatação existente
    phone = re.sub(r"[^0-9]", "", phone)

    if len(phone) == 11:  # Celular
        return f"({phone[:2]}) {phone[2:7]}-{phone[7:]}"
    elif len(phone) == 10:  # Fixo
        return f"({phone[:2]}) {phone[2:6]}-{phone[6:]}"

    return phone


def format_cep(cep: str) -> str:
    """Formata CEP para exibição.

    Args:
        cep: CEP sem formatação

    Returns:
        str: CEP formatado (XXXXX-XXX)
    """
    if not cep:
        return ""

    # Remove formatação existente
    cep = re.sub(r"[^0-9]", "", cep)

    if len(cep) == 8:
        return f"{cep[:5]}-{cep[5:]}"

    return cep
